File: codewars/4-kyu/sum-of-intervals/solution.rs
// https://www.codewars.com/kata/52b7ed099cdc285c300001cd

// solution 1: 2130ms
extern crate alloc;
pub fn sum_intervals(intervals: &[(i32, i32)]) -> i32 {
    let mut intervals = intervals.to_vec();
    intervals.sort_unstable_by_key(|&(a, _)| a);

    let mut res = 0;
    let mut end = i32::MIN;
    for (a, b) in intervals {
        if end < a {
            end = a;
        }
        if end < b {
            res += b - end;
            end = b;
        }
    }
    res
}

// solution 2: 1560ms
use itertools::Itertools;
fn sum_intervals(intervals: &[(i32, i32)]) -> i32 {
    let mut sum = 0;
    let mut end = i32::MIN;
    for (a, b) in intervals.iter().sorted() {
        if end < *a {
            end = *a;
        }
        if end < *b {
            sum += b - end;
            end = *b;
        }
    }
    sum
}

File: codewars/4-kyu/sum-of-intervals/solution.py
# https://www.codewars.com/kata/52b7ed099cdc285c300001cd

# solution 1: 593ms
def sum_of_intervals(intervals: list) -> int:
    total=0
    intervals=sorted(intervals)
    a=intervals[0][0]
    b=intervals[0][1]
    total += b - a
    if len(set(intervals)) == 1:
        return total
    else:
        for j in intervals[1:]:
            if j[0] > b:
                total += j[1] - j[0]
                a = j[0]
                b = j[1]
            elif j[0] >= a and j[1] <= b:
                continue
            else:
                total += j[1] - b
                b=j[1]
    return total


# solution 2: 588ms
def sum_of_intervals(intervals):
    intervals.sort()
    end,subtract = intervals[0][1],0
    for i in intervals:
        if i[0] > end: subtract += i[0]-end
        if i[1] > end: end = i[1]
    return end - subtract - intervals[0][0]


# solution 3: 547ms
from bisect import bisect 
def sum_of_intervals(intervals):
    I = []
    for a,b in intervals:
        i = bisect(I, a)
        j = bisect(I, b)
        I = I[:i] + [a]*(1-i%2) + [b]*(1-j%2) + I[j:]
    
    return sum(I[1::2]) - sum(I[::2])